libethdrivers/src/plat/imx6/imx6.c

```diff
@@ -348,32 +325,38 @@ static void complete_rx(imx6_eth_driver_t *dev)
     ethif_raw_rx_complete cb_complete = dev->eth_drv.i_cb.rx_complete;
     assert(cb_complete);
 
-    unsigned int rdt = dev->rdt;
-    while (dev->rdh != rdt) {
-        unsigned int status = dev->rx_ring[dev->rdh].stat;
-        /* Ensure no memory references get ordered before we checked the
-         * descriptor was written back
+    ring_ctx_t *ring = &(dev->rx);
+    unsigned int head = ring->head;
+
+    /* Release all descriptors that have data. */
+    while (head != ring->tail) {
+
+        /* The NIC hardware can modify the descriptor any time, 'volatile'
+         * prevents the compiler's optimizer from caching values and enforces
+         * every access happen as stated in the code.
          */
-        __sync_synchronize();
-        if (status & RXD_EMPTY) {
-            /* not complete yet */
+        volatile struct descriptor *d = &(ring->descr[head]);
+
+        /* If the slot is still marked as empty we are done. */
+        if (d->stat & RXD_EMPTY) {
+            assert(dev->enet);
+            if (!enet_rx_enabled(dev->enet)) {
+                enet_rx_enable(dev->enet);
+            }
             break;
         }
-        void *cookie = dev->rx_cookies[dev->rdh];
-        unsigned int len = dev->rx_ring[dev->rdh].len;
-        /* update rdh */
-        dev->rdh = (dev->rdh + 1) % dev->rx_size;
-        dev->rx_remain++;
-        /* Give the buffers back */
-        cb_complete(cb_cookie, 1, &cookie, &len);
-    }
 
-    if (dev->rdt != dev->rdh) {
-        struct enet *enet = dev->enet;
-        assert(enet);
-        if (!enet_rx_enabled(enet)) {
-            enet_rx_enable(enet);
+        void *cookie = ring->cookies[head];
+        /* Go to next buffer, handle roll-over. */
+        if (++head == ring->cnt) {
+            head = 0;
         }
+        ring->head = head;
+        /* if rx and tx happens in parallel, this can have a race condition */
```

Comment:
I don't see how the race condition can happen when the ring data structures are independent of each other. Am I missing something here?

Reply:
I has hoping for some discussion here of there could be an issue or not. Basically, the queue could be use by two entities, a producer and a consumer. So the question is, how thread-safe it is or if the access if serialized somewhere else. The comment that that rx/tx happens in parallel is misleading, it's actually the producer and consumer running in parallel.

Reply:
Right. This isn't thread-safe at all and access to this isn't serialized at all. But we've been getting along by using single-threaded environments and making sure that access is serialized.

Reply:
In fact, all libraries in util_libs relies on this assumption.

Reply:
I'll improve the comment to state this

Reply:
done